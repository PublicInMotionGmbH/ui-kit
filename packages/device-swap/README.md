# Talixo Device Swap

UI Component which represents Device Swap

## How to install

Package is available as `@talixo/mobile-swap` in NPM registry, so you can use it in your project
using `npm install @talixo/mobile-swap --save` or `yarn add @talixo/mobile-swap`.

## Requirements

Your package should additionally have some extra dependencies:

- `prop-types: ^15.6.1`
- `react: ^16.6.3`
- `react-dom: ^16.6.3`

These packages are required by `@talixo/mobile-swap`, but you have to install them manually,
to avoid having different versions of these in your application.

## Supported props

There are 3 properties used directly by this component, all other are passed to renderers.

Property name | Type      | Default     | Description                    
--------------|-----------|:-----------:|--------------------------------
defaultView   | string    | `"desktop"` | `"desktop"` or `"mobile"` - view used until device type is determined
renderMobile  | function  | n/a         | renderer for mobile view
renderDesktop | function  | n/a         | renderer for desktop view

## Changelog

- **0.1.0** - initial version
