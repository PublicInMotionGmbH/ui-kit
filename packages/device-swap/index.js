import DeviceSwap from './src/DeviceSwap'
import DeviceDetector from './src/DeviceDetector'
import detector from './src/detector'

export {
  DeviceSwap,
  DeviceDetector,
  detector
}
