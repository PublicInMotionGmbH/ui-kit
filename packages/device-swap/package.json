{
  "name": "@talixo/device-swap",
  "version": "1.0.0-alpha.31",
  "description": "UI Component which helps to render different components on different devices",
  "main": "dist/index.js",
  "module": "index.js",
  "jsnext:main": "index.js",
  "scripts": {
    "test": "node ../../scripts/runTests.js --only device-swap",
    "prepublishOnly": "npm run build",
    "build": "node ../../scripts/buildPackages.js --only device-swap"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Public in Motion GmbH",
  "repository": {
    "type": "git",
    "url": "https://github.com/PublicInMotionGmbH/ui-kit.git"
  },
  "bugs": {
    "url": "https://github.com/PublicInMotionGmbH/ui-kit/issues"
  },
  "license": "MIT",
  "devDependencies": {
    "@talixo/shared": "^1.0.0-alpha.31",
    "prop-types": "^15.6.1",
    "react": "^16.2.0",
    "react-dom": "^16.2.0"
  },
  "peerDependencies": {
    "prop-types": "^15.6.1",
    "react": "^16.2.0",
    "react-dom": "^16.2.0"
  }
}
