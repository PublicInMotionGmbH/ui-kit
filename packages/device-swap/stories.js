import React from 'react'
import { createStoriesFactory, getReadmeDescription } from '@talixo/shared/story'

import DeviceSwap from './src/DeviceSwap'

// Load first paragraph from README file
const readme = getReadmeDescription(require('./README.md'))

// Create factories for story
const addStory = createStoriesFactory('Device Swap', module, {
  propTables: [ DeviceSwap ]
})

// Stories

addStory('initial', readme, () => (
  <DeviceSwap
    renderDesktop={() => 'Desktop'}
    renderMobile={() => 'Mobile'}
  />
))

addStory('default view: mobile', readme, () => (
  <DeviceSwap
    defaultView='mobile'
    renderDesktop={() => 'Desktop'}
    renderMobile={() => 'Mobile'}
  />
))
