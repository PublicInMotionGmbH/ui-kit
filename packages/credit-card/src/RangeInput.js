import React from 'react'
import PropTypes from 'prop-types'

import { buildClassName } from '@talixo/shared'

import { AutoComplete, SelectBox } from '@talixo/combo-box'
import { TextInput } from '@talixo/text-input'
import { Icon } from '@talixo/icon'
import { DeviceSwap } from '@talixo/device-swap'

const moduleName = 'range-input'

const propTypes = {

  /** Minimum input value length. */
  minLength: PropTypes.number,

  /** Additional class name. */
  className: PropTypes.string,

  /** Minimum value. */
  min: PropTypes.number,

  /** Maximum value. */
  max: PropTypes.number,

  /** Callback for blur event. */
  onBlur: PropTypes.func,

  /** Callback for focus event. */
  onFocus: PropTypes.func,

  /** Callback for change event. */
  onChange: PropTypes.func,

  /** Passed value. */
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),

  /** Input value length for auto complete. */
  autoCompleteLength: PropTypes.number,

  /** Should it render native select box on mobile? */
  mobileFriendly: PropTypes.number
}

const defaultProps = {
  minLength: 1,
  mobileFriendly: false
}

/**
 * Format value.
 *
 * @param {number} value
 * @param {number} minLength
 *
 * @returns {string}
 */
function formatValue (value, minLength) {
  if (value == null) {
    return ''
  }

  value = +value

  if (isNaN(value)) {
    return ''
  }

  value = value.toString()

  while (value.length < minLength) {
    value = '0' + value
  }

  return value
}

/**
 * Build end value.
 *
 * @param {string} value
 * @param {number} min
 * @param {number} max
 *
 * @returns {number}
 */
function buildEndValue (value, min, max) {
  if (value !== '0' && value !== 0 && !value) {
    return null
  }

  value = +value

  if (isNaN(value)) {
    return null
  }

  if (value > max) {
    return null
  }

  if (value < min) {
    return null
  }

  return value
}

/**
 * Create an array of numbers in desired range.
 *
 * @param {number} min
 * @param {number} max
 *
 * @returns {array}
 */
function range (min, max) {
  const arr = []

  for (let i = min; i <= max; i++) {
    arr.push(i)
  }

  return arr
}

/**
 * Component which represents expiration date input.
 *
 * @property {object} props
 * @property {number} [props.minLength]
 * @property {string} [props.className]
 * @property {number} [props.min]
 * @property {number} [props.max]
 * @property {func} [props.onBlur]
 * @property {func} [props.onFocus]
 * @property {func} [props.onChange]
 * @property {object} [props.value]
 * @property {number} [props.value.month]
 * @property {number} [props.value.year]
 * @property {number} [props.autoCompleteLength]
 *
 * @property {object} state
 * @property {object|null} state.inputValue
 * @property {object|null} state.value
 * @property {object|null} state.month
 * @property {object|null} state.year
 *
 * @class
 */
class RangeInput extends React.PureComponent {
  state = {
    inputValue: formatValue(this.props.value, this.props.minLength),
    value: this.props.value || null,
    focus: false,
    options: range(this.props.min, this.props.max)
  }

  componentWillReceiveProps (props) {
    if (props.value !== this.props.value && props.value !== undefined) {
      this.setState({
        value: props.value
      })
    }

    if (!this.state.focus && props.value !== undefined && (props.value !== this.state.inputValue || props.minLength !== this.props.minLength)) {
      this.setState({
        inputValue: formatValue(props.value, props.minLength)
      })
    }

    if (props.min !== this.props.min || props.max !== this.props.max) {
      this.setState({
        options: range(props.min, props.max)
      })
    }
  }

  /**
   * Handle input value change.
   *
   * @param {string} value
   */
  onInputValueChange = (value) => {
    // If passed value is the same as in current state, return.
    if (value === this.state.inputValue) {
      return
    }

    this.setState({ inputValue: value })
  }

  /**
   * Handle change.
   *
   * @param {string} value
   */
  onChange = (value) => {
    const { onChange } = this.props

    // Build value
    value = buildEndValue(value, this.props.min, this.props.max)

    if (this.props.value === undefined) {
      this.setState({
        value,
        inputValue: formatValue(value, this.props.minLength)
      })
    }

    if (onChange) {
      onChange(value)
    }

    if (this.node) {
      this.node.querySelector('input').blur()
    }
  }

  /**
   * Handle focus.
   *
   */
  focus = () => {
    const { onFocus } = this.props

    this.setState({
      focus: true
    })

    if (onFocus) {
      onFocus()
    }
  }

  /**
   * Handle blur.
   *
   */
  blur = () => {
    const { autoCompleteLength, onBlur, onChange } = this.props
    const { inputValue } = this.state

    const state = { focus: false }

    if (onBlur) {
      onBlur()
    }

    let value = buildEndValue(inputValue, this.props.min, this.props.max)

    if (value === null && autoCompleteLength && autoCompleteLength <= inputValue.length) {
      const options = this.state.options.filter(x => `${x}`.substr(-inputValue.length) === inputValue)

      if (options.length === 1) {
        value = options[0]
      }
    }

    if (inputValue) {
      state.inputValue = formatValue(value, this.props.minLength)
    }

    this.setState(state)

    if (this.props.value === undefined) {
      this.setState({ value })
    }

    if (this.props.value === value) {
      return
    }

    if (onChange) {
      onChange(value)
    }
  }

  /**
   * Handle key down.
   *
   * @param {SyntheticEvent} event
   */
  onKeyDown = (event) => {
    const charCode = event.which || event.keyCode

    if (charCode > 31 && (charCode < 48 || charCode > 57) && (charCode < 37 || charCode > 40)) {
      event.preventDefault()
    }
  }

  /**
  * Save base node element.
  *
  * @param {Element} node
  */
  saveRef = (node) => {
    this.node = node
  }

  renderComboBox = () => {
    const {
      minLength, className, min, max, onBlur, onFocus, onChange,
      value: _value, mobileFriendly, autoCompleteLength, ...passedProps
    } = this.props
    const { focus, options, value, inputValue } = this.state

    // Build icon.
    const icon = this.state.focus
      ? <Icon name='keyboard_arrow_up' />
      : <Icon name='keyboard_arrow_down' />

    return (
      <AutoComplete
        options={options}
        isOpen={focus}
        renderItem={(item) => formatValue(item, minLength)}
        value={value}
        inputValue={inputValue}
        onInputValueChange={this.onInputValueChange}
        onChange={this.onChange}
        onFocus={this.focus}
        onBlur={this.blur}
        itemToString={x => x == null ? '' : formatValue(x, minLength)}
      >
        <TextInput
          onKeyDown={this.onKeyDown}
          pattern='[0-9]*'
          inputMode='numeric'
          right={icon}
          maxLength={Math.max(('' + max).length, minLength)}
          {...passedProps}
        />
      </AutoComplete>
    )
  }

  renderSelectBox = () => {
    const {
      minLength, className, min, max, onBlur, onFocus, onChange,
      value: _value, autoCompleteLength, mobileFriendly, ...passedProps
    } = this.props
    const { focus, options, value } = this.state

    return (
      <SelectBox
        {...passedProps}
        mobileFriendly
        options={options}
        isOpen={focus}
        renderItem={item => formatValue(item, minLength)}
        value={value}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
      />
    )
  }

  /**
   * Render range input component.
   *
   * @returns {React.Element}
   */
  render () {
    const { className, mobileFriendly } = this.props
    const { focus } = this.state

    // Build class names.
    const wrapperClsName = buildClassName(moduleName, className, {
      focused: focus
    })

    const inner = mobileFriendly ? (
      <DeviceSwap
        defaultView='mobile'
        renderMobile={this.renderSelectBox}
        renderDesktop={this.renderComboBox}
      />
    ) : this.renderComboBox()

    return (
      <div className={wrapperClsName} ref={this.saveRef}>
        {inner}
      </div>
    )
  }
}

RangeInput.propTypes = propTypes

RangeInput.defaultProps = defaultProps

export default RangeInput
