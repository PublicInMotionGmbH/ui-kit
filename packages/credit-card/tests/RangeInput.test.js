import React from 'react'
import { mount, shallow } from 'enzyme'

import { detector } from '@talixo/device-swap'

import RangeInput from '../src/RangeInput'

function createKeyEvent (keyCode, fn) {
  return {
    keyCode: keyCode,
    preventDefault: fn
  }
}

describe('<RangeInput />', () => {
  beforeEach(() => detector.setViewType('desktop'))
  afterEach(() => detector.reset())

  it('renders correctly', () => {
    const wrapper = shallow(<RangeInput />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly mounted', () => {
    const wrapper = mount(<RangeInput />)

    expect(wrapper).toMatchSnapshot()
    wrapper.unmount()
  })

  it('updates state.value when value is provided', () => {
    const value = 3
    const wrapper = shallow(<RangeInput />)
    wrapper.setProps({ value })
    expect(wrapper.state().value).toEqual(value)
  })
})

describe('formatValue', () => {
  it('formats value correctly if minLength is provided', () => {
    const value = 1
    const minLength = 3

    const wrapper = shallow(<RangeInput value={value} minLength={minLength} />)
    expect(wrapper.state().inputValue).toEqual('001')
  })

  it('formats value correctly if NaN is provided', () => {
    const value = NaN

    const wrapper = shallow(<RangeInput value={value} />)
    expect(wrapper.state().inputValue).toEqual('')
  })
})

describe('componentWillReceiveProps', () => {
  let value, minLength, wrapper
  beforeEach(() => {
    value = 1
    minLength = 3

    wrapper = shallow(<RangeInput value={value} minLength={minLength} />)
    wrapper.state.focus = false
  })

  it('sets inputValue correctly if value is passed', () => {
    wrapper.setProps({ value: 2 })
    expect(wrapper.state().inputValue).toEqual('002')
  })

  it('sets inputValue correctly if minLength is passed', () => {
    wrapper.setProps({ minLength: 2 })
    expect(wrapper.state().inputValue).toEqual('01')
  })
})

describe('buildEndValue', () => {
  it('builds end value correctly', () => {
    const wrapper = shallow(<RangeInput />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change', '1')
    expect(wrapper.state().value).toEqual(1)
  })

  it('returns null if value is greater than max', () => {
    const min = 1
    const max = 12

    const wrapper = shallow(<RangeInput min={min} max={max} />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change', '14')
    expect(wrapper.state().value).toEqual(null)
  })

  it('returns null if value is smaller than min', () => {
    const min = 7
    const max = 12

    const wrapper = shallow(<RangeInput min={min} max={max} />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change', '1')
    expect(wrapper.state().value).toEqual(null)
  })

  it('returns null if value is NaN', () => {
    const wrapper = shallow(<RangeInput />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change', 'string')
    expect(wrapper.state().value).toEqual(null)
  })

  it('returns null if value is false', () => {
    const wrapper = shallow(<RangeInput />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change', false)
    expect(wrapper.state().value).toEqual(null)
  })

  it('updates state.inputValue on input value change', () => {
    const inputValue = '15'
    const wrapper = shallow(<RangeInput />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change', inputValue)
    expect(wrapper.state().inputValue).toEqual('15')
  })
})

describe('range', () => {
  it('formats state.options correctly', () => {
    const min = 1
    const max = 10
    const range = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

    const wrapper = shallow(<RangeInput min={min} max={max} />)

    expect(wrapper.state().options).toEqual(range)
  })

  it('formats state.options correctly if min is provided', () => {
    const min = 1
    const max = 10
    const range = [5, 6, 7, 8, 9, 10]

    const wrapper = shallow(<RangeInput min={min} max={max} />)
    wrapper.setProps({ min: 5 })

    expect(wrapper.state().options).toEqual(range)
  })

  it('formats state.options correctly if max is provided', () => {
    const min = 1
    const max = 10
    const range = [1, 2, 3, 4, 5]

    const wrapper = shallow(<RangeInput min={min} max={max} />)
    wrapper.setProps({ max: 5 })

    expect(wrapper.state().options).toEqual(range)
  })
})

describe('change', () => {
  it('calls onChange', () => {
    const change = jest.fn()
    const wrapper = shallow(<RangeInput onChange={change} />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change')

    expect(change).toHaveBeenCalledTimes(1)
  })
})

describe('focus', () => {
  it('updates state.focus', () => {
    const wrapper = shallow(<RangeInput />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('focus')

    expect(wrapper.state().focus).toEqual(true)
  })

  it('calls onFocus', () => {
    const focus = jest.fn()
    const wrapper = shallow(<RangeInput onFocus={focus} />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('focus')

    expect(focus).toHaveBeenCalledTimes(1)
  })
})

describe('blur', () => {
  it('updates state.focus', () => {
    const wrapper = shallow(<RangeInput />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('focus')
    autoComplete.simulate('blur')

    expect(wrapper.state().focus).toEqual(false)
  })

  it('calls onBlur', () => {
    const blur = jest.fn()
    const wrapper = shallow(<RangeInput onBlur={blur} />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('blur')

    expect(blur).toHaveBeenCalledTimes(1)
  })

  it('formats inputValue', () => {
    const wrapper = shallow(<RangeInput minLength={3} />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('change', 12)
    autoComplete.simulate('blur')

    expect(wrapper.state().inputValue).toEqual('012')
  })

  it('calls onChange', () => {
    const change = jest.fn()
    const wrapper = shallow(<RangeInput onChange={change} />)
    const autoComplete = wrapper.find('AutoComplete')
    autoComplete.simulate('blur')

    expect(change).toHaveBeenCalledTimes(1)
  })

  it('doesn`t call onChange if value is the same as in props', () => {
    const change = jest.fn()
    const value = 12
    const wrapper = shallow(<RangeInput onChange={change} value={value} />)
    const autoComplete = wrapper.find('AutoComplete')

    autoComplete.simulate('change', '12')
    jest.resetAllMocks()
    autoComplete.simulate('blur')

    expect(change).toHaveBeenCalledTimes(0)
  })

  it('autocompletes value correctly', () => {
    const autoCompleteLength = 2
    const min = 2000
    const max = 2020

    const wrapper = shallow(<RangeInput
      autoCompleteLength={autoCompleteLength}
      min={min}
      max={max}
    />)
    const autoComplete = wrapper.find('AutoComplete')
    wrapper.setState({ inputValue: '12' })
    autoComplete.simulate('blur')

    expect(wrapper.state().value).toEqual(2012)
  })
})

describe('onKeyDown', () => {
  const event = {
    preventDefault: () => {}
  }
  let wrapper, textInput

  beforeEach(() => {
    jest.spyOn(event, 'preventDefault')
    wrapper = shallow(<RangeInput />)
    console.log(wrapper.children())
    textInput = wrapper.find('TextInput')
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('doesn`t prevent default on digit and arrow keys', () => {
    textInput.simulate('keydown', createKeyEvent(38, event.preventDefault))
    expect(event.preventDefault).toHaveBeenCalledTimes(0)
    textInput.simulate('keydown', createKeyEvent(49, event.preventDefault))
    expect(event.preventDefault).toHaveBeenCalledTimes(0)
  })

  it('prevents default on the rest', () => {
    textInput.simulate('keydown', createKeyEvent(32, event.preventDefault))
    expect(event.preventDefault).toHaveBeenCalledTimes(1)

    jest.resetAllMocks()

    textInput.simulate('keydown', createKeyEvent(58, event.preventDefault))
    expect(event.preventDefault).toHaveBeenCalledTimes(1)
  })
})

describe('onInputValueChange', () => {
  it('sets value as state.inputValue', () => {
    const wrapper = shallow(<RangeInput />)
    const autoComplete = wrapper.find('AutoComplete')

    autoComplete.props().onInputValueChange('12')
    expect(wrapper.state().inputValue).toEqual('12')
  })

  it('returns is value is the same as in state.inputValue', () => {
    const wrapper = shallow(<RangeInput value={12} />)
    const autoComplete = wrapper.find('AutoComplete')

    autoComplete.props().onInputValueChange('12')
    expect(wrapper.state().inputValue).toEqual('12')
  })
})
