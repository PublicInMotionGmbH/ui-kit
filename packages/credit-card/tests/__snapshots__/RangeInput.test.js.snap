// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeInput /> renders correctly 1`] = `
<div
  className="talixo-range-input"
>
  <AutoComplete
    buildItemId={[Function]}
    inputValue=""
    isOpen={false}
    itemToString={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onInputValueChange={[Function]}
    options={Array []}
    renderItem={[Function]}
    value={null}
  >
    <TextInput
      InputComponent="input"
      error={false}
      inputMode="numeric"
      maxLength={9}
      onKeyDown={[Function]}
      pattern="[0-9]*"
      right={
        <Icon
          name="keyboard_arrow_down"
        />
      }
      type="text"
    />
  </AutoComplete>
</div>
`;

exports[`<RangeInput /> renders correctly mounted 1`] = `
<div
  className="talixo-range-input"
>
  <div
    className="talixo-combo-box"
  >
    <div
      className="talixo-text-input talixo-text-input--with-right talixo-text-input--empty"
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="talixo-text-input__input"
        id="downshift-0-input"
        inputMode="numeric"
        maxLength={9}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        pattern="[0-9]*"
        role="combobox"
        style={
          Object {
            "paddingLeft": 1,
            "paddingRight": 1,
          }
        }
        type="text"
        value=""
      />
      <span
        className="talixo-text-input__right"
      >
        <span
          aria-hidden="true"
          className="talixo-icon tai tai--material"
        >
          îŒ“
        </span>
      </span>
    </div>
  </div>
</div>
`;
